### PR TITLE
[WIP] Apply to collection optional dtype

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Makes dtype optional in `apply_to_collection`. This allows filtering inside the `function`.

Draft until #5954 is merged

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -449,7 +449,7 @@ def all_gather(
 
         data = convert_to_tensors(data, device=self.device)
         all_gather = partial(all_gather, group=group, sync_grads=sync_grads)
-        return apply_to_collection(data, torch.Tensor, all_gather)
+        return apply_to_collection(data, all_gather, dtype=torch.Tensor)
```
Comment (tchaton) [inline pytorch_lightning/core/lightning.py:452]: Any reasons for switching them ? It is breaking BC.

**Comment by Borda** [comment]
@carmocca how is it going here? 🐰

**Comment by justusschock** [comment]
@carmocca is this really necessary? You should be able to just specify dtype=object and achieve the same result and it also does not save any checks.

**Comment by carmocca** [comment]
> @carmocca is this really necessary? You should be able to just specify dtype=object and achieve the same result and it also does not save any checks.

Yeah, I guess that works too. I just thought it would be clearner to have dtype be optional but IDK how we could do it without breaking BC.

File: pytorch_lightning/core/lightning.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/core/lightning.py:452]: Can't have keyword arguments before positional

Comment (Borda) [reply pytorch_lightning/core/lightning.py:452]: then just call dem as `argument=value`

Comment (carmocca) [reply pytorch_lightning/core/lightning.py:452]: The issue is not the order here but in the signature

**Pull request closed without merging**